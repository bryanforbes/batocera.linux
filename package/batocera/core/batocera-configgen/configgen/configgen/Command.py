from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Mapping, Sequence
    from pathlib import Path


class Command:
    def __init__(self, array: Sequence[str | Path], env: Mapping[str, str | Path] = {}):
        self.array = list(array)
        self.env = dict(env)

    def __str__(self):
        strings = [f"{varName}={varValue}" for varName, varValue in self.env.items()]

        strings.extend(str(value) for value in self.array)

        return " ".join(strings)
