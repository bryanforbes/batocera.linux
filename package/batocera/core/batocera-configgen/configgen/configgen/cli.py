from __future__ import annotations

import logging
import time
from collections.abc import Callable
from functools import partial
from pathlib import Path
from typing import Any, TypedDict, TypeVar, Unpack, cast

import click

from . import profiler
from .emulators import run_emulator
from .utils.logger import setup_logging

_FC = TypeVar('_FC', bound=Callable[..., Any] | click.Command)

_logger = logging.getLogger(__name__)


def _player_callback(player: int, ctx: click.Context, option: click.Option, value: int | None) -> int | None:
    current_value: list[PlayerOptions] = ctx.params.setdefault('players', [])

    option_names = (
        f'p{player}hat_count',
        f'p{player}button_count',
        f'p{player}device_path',
        f'p{player}real_name',
        f'p{player}guid',
        f'p{player}index',
    )

    items = [
        (f'p{player}axis_count', value),
        *[(option_name, ctx.params.pop(option_name)) for option_name in option_names],
    ]

    if all(item[1] is not None for item in items):
        p_length = len(f'p{player}')
        current_value.append(
            cast(
                PlayerOptions,
                {
                    'player_number': player,
                    **{option_name[p_length:]: option_value for option_name, option_value in items},
                },
            )
        )
        return value

    if all(item[1] is None for item in items):
        return value

    raise click.UsageError(f'All options that start with p{player} must be specified if any are specified', ctx=ctx)


def _players(max_players: int) -> Callable[[_FC], _FC]:
    def wrap(callable: _FC) -> _FC:
        for n in range(max_players, 0, -1):
            callable = click.option(f'-p{n}nbaxes', f'p{n}axis_count', type=int, callback=partial(_player_callback, n))(
                callable
            )
            callable = click.option(f'-p{n}nbhats', f'p{n}hat_count', type=int)(callable)
            callable = click.option(f'-p{n}nbbuttons', f'p{n}button_count', type=int)(callable)
            callable = click.option(f'-p{n}devicepath', f'p{n}device_path', type=str)(callable)
            callable = click.option(f'-p{n}name', f'p{n}real_name', type=str)(callable)
            callable = click.option(f'-p{n}guid', type=str)(callable)
            callable = click.option(f'-p{n}index', type=int)(callable)

        return callable

    return wrap


class PlayerOptions(TypedDict):
    player_number: int
    index: int
    guid: str
    real_name: str
    device_path: str
    button_count: int
    hat_count: int
    axis_count: int


class Options(TypedDict):
    system: str
    rom: Path
    emulator: str | None
    core: str | None
    netplaymode: str | None
    netplaypass: str | None
    netplayip: str | None
    netplayport: str | None
    netplaysession: str | None
    state_slot: str | None
    state_filename: Path | None
    autosave: str | None
    systemname: str | None
    gameinfoxml: Path
    lightgun: bool
    wheel: bool
    trackball: bool
    spinner: bool
    players: list[PlayerOptions]


@click.command()
@click.option('-system', required=True)
@click.option('-rom', required=True, type=click.Path(path_type=Path))
@click.option('-emulator')
@click.option('-core')
@click.option('-netplaymode')
@click.option('-netplaypass')
@click.option('-netplayip')
@click.option('-netplayport')
@click.option('-netplaysession')
@click.option('-state_slot')
@click.option('-state_filename', type=click.Path(path_type=Path))
@click.option('-autosave')
@click.option('-systemname')
@click.option('-gameinfoxml', default='/dev/null', type=click.Path(path_type=Path))
@click.option('-lightgun', is_flag=True)
@click.option('-wheel', is_flag=True)
@click.option('-trackball', is_flag=True)
@click.option('-spinner', is_flag=True)
@_players(8)
def cli(**kwargs: Unpack[Options]) -> None:
    with setup_logging():
        exit_code = -1

        try:
            exit_code = run_emulator(kwargs)
        except Exception:
            _logger.exception('configgen exception: ')

        profiler.disable(dump=True)

        time.sleep(1)  # this seems to be required so that the gpu memory is restituated and available for es
        _logger.debug('Exiting configgen with status %s', exit_code)

        exit(exit_code)
