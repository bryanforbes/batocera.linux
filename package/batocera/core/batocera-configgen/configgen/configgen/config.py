from __future__ import annotations

import logging
import xml.etree.ElementTree as ET
from collections.abc import Iterator, Mapping
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any, ClassVar, Literal, Self, TypeVar, overload

import yaml

from .batoceraPaths import BATOCERA_SHADERS, DEFAULTS_DIR, ES_SETTINGS, USER_SHADERS
from .utils.missing import MISSING, MissingType

if TYPE_CHECKING:
    from pathlib import Path

    from .cli import Options
    from .settings.unixSettings import UnixSettings

_T = TypeVar('_T')
_TV = TypeVar('_TV')
_FV = TypeVar('_FV')

_logger = logging.getLogger(__name__)


# adapted from https://gist.github.com/angstwad/bf22d1822c38a92ec0a9
def _dict_merge(destination: dict[str, Any], source: Mapping[str, Any]) -> None:
    """Recursive dict merge. Inspired by :meth:``dict.update()``, instead of
    updating only top-level keys, dict_merge recurses down into dicts nested
    to an arbitrary depth, updating keys. The ``merge_dct`` is merged into
    ``dct``.
    :param destination: dict onto which the merge is executed
    :param source: dict merged into destination
    :return: None
    """
    for key, value in source.items():
        if key in destination and isinstance(destination[key], dict) and isinstance(value, Mapping):
            _dict_merge(destination[key], value)
        else:
            destination[key] = value


def _load_defaults(system: str, default_file: Path, default_arch_file: Path, /) -> dict[str, Any]:
    with default_file.open() as f:
        defaults: dict[str, Any] = yaml.load(f, Loader=yaml.CLoader)

    with default_arch_file.open() as f:
        arch_defaults: dict[str, Any] = yaml.load(f, Loader=yaml.CLoader) or {}

    config: dict[str, Any] = defaults.get('default', {})

    if 'default' in arch_defaults:
        _dict_merge(config, arch_defaults['default'])

    if system in defaults:
        _dict_merge(config, defaults[system])

    if system in arch_defaults:
        _dict_merge(config, arch_defaults[system])

    return config


@dataclass(slots=True)
class Config:
    TRUE_VALUES: ClassVar[set[Literal['1', 'true', 'on', 'enabled', True]]] = {'1', 'true', 'on', 'enabled', True}
    MISSING: ClassVar = MISSING

    data: dict[str, Any]

    def __contains__(self, x: object, /) -> bool:
        return x in self.data

    @overload
    def get(self, key: str, default: MissingType = ..., /) -> Any | MissingType: ...

    @overload
    def get(self, key: str, default: _T, /) -> Any | _T: ...

    def get(self, key: str, default: _T | MissingType = MISSING, /) -> Any | _T | MissingType:
        return self.data.get(key, default)

    @overload
    def get_bool(self, key: str, default: bool = False, /, *, return_values: None = None) -> bool: ...

    @overload
    def get_bool(self, key: str, default: bool = False, /, *, return_values: tuple[_TV, _FV]) -> _TV | _FV: ...

    def get_bool(
        self, key: str, default: bool = False, /, *, return_values: tuple[_TV, _FV] | None = None
    ) -> bool | _TV | _FV:
        value = self.get(key)

        if value is MISSING:
            return default

        if isinstance(value, str):
            value = value.lower()

        if return_values is None:
            return value in self.TRUE_VALUES

        return return_values[value not in self.TRUE_VALUES]

    @overload
    def get_str(self, key: str, default: MissingType = ..., /) -> str | MissingType: ...

    @overload
    def get_str(self, key: str, default: str, /) -> str: ...

    def get_str(self, key: str, default: str | MissingType = MISSING, /) -> str | MissingType:
        value = self.get(key)

        if value is MISSING:
            return default

        return str(value)

    @overload
    def get_int(self, key: str, default: MissingType = ..., /) -> int | MissingType: ...

    @overload
    def get_int(self, key: str, default: int, /) -> int: ...

    def get_int(self, key: str, default: int | MissingType = MISSING, /) -> int | MissingType:
        value = self.get(key)

        if value is MISSING:
            return default

        return int(value)

    @overload
    def get_float(self, key: str, default: MissingType = ..., /) -> float | MissingType: ...

    @overload
    def get_float(self, key: str, default: float, /) -> float: ...

    def get_float(self, key: str, default: float | MissingType = MISSING, /) -> float | MissingType:
        value = self.get(key)

        if value is MISSING:
            return default

        return float(value)

    def items(self, /, *, starts_with: str | None = None) -> Iterator[tuple[str, Any]]:
        if starts_with is None:
            yield from self.data.items()
        else:
            starts_with_len = len(starts_with)
            for key, value in self.data.items():
                if key.startswith(starts_with):
                    yield key[starts_with_len:], value


def _get_rom_settings_name(rom: Path, /) -> str:
    # sanitize rule by EmulationStation
    # see FileData::getConfigurationName() on batocera-emulationstation
    rom_settings_name = rom.name.replace('=', '').replace('#', '')

    _logger.info('game settings name: %s', rom_settings_name)

    return rom_settings_name


@dataclass
class SystemConfig(Config):
    @property
    def core(self) -> str:
        return self.data['core']

    @property
    def core_forced(self) -> str:
        return self.data['core-forced']

    @property
    def emulator(self) -> str:
        return self.data['emulator']

    @property
    def emulator_forced(self) -> str:
        return self.data['emulator-forced']

    @property
    def ui_mode(self) -> Literal['Full', 'Kiosk', 'Kid']:
        return self.data['uimode']

    @property
    def show_fps(self) -> bool:
        return self.data['showFPS']

    @property
    def video_mode(self) -> str:
        return self.data.get('videomode', 'default')

    @property
    def use_guns(self) -> bool:
        return self.get_bool('use_guns')

    @property
    def use_wheels(self) -> bool:
        return self.get_bool('use_wheels')

    @classmethod
    def load(cls, options: Options, settings: UnixSettings, /) -> Self:
        defaults = _load_defaults(
            options['system'], DEFAULTS_DIR / 'configgen-defaults.yml', DEFAULTS_DIR / 'configgen-defaults-arch.yml'
        )
        data: dict[str, Any] = {
            'emulator': defaults['emulator'],
            'core': defaults['core'],
        }

        # In the yaml files, the "options" structure is not flat, so we have to flatten it here
        # because the options are flat in batocera.conf to make it easier for end users to edit
        if 'options' in defaults:
            _dict_merge(data, defaults['options'])

        rom = options['rom']
        rom_settings_name = _get_rom_settings_name(rom)

        data.update(settings.get_all_iter('display', keep_name=True, keep_defaults=True))
        data.update(settings.get_all_iter('controllers', keep_name=True))

        global_settings = settings.get_all('global')
        system_settings = settings.get_all(options['system'])
        folder_settings = settings.get_all(f'{options["system"]}.folder["{rom.parent}"]')
        game_settings = settings.get_all(f'{options["system"]}["{rom_settings_name}"]')

        data.update(global_settings)
        data.update(system_settings)
        data.update(folder_settings)
        data.update(game_settings)

        try:
            es_config = ET.parse(ES_SETTINGS)

            # showFPS
            drawframerate_node = es_config.find("./bool[@name='DrawFramerate']")
            drawframerate_value = drawframerate_node.attrib['value'] if drawframerate_node is not None else 'false'
            if drawframerate_value not in ['false', 'true']:
                drawframerate_value = 'false'

            data['showFPS'] = drawframerate_value == 'true'

            # uimode
            uimode_node = es_config.find("./string[@name='UIMode']")
            uimode_value = uimode_node.attrib['value'] if uimode_node is not None else 'Full'
            if uimode_value not in ['Full', 'Kiosk', 'Kid']:
                uimode_value = 'Full'

            data['uimode'] = uimode_value
        except Exception:
            data['showFPS'] = False
            data['uimode'] = 'Full'

        data['emulator-forced'] = (
            'emulator' in global_settings
            or 'emulator' in system_settings
            or 'emulator' in game_settings
            or options['emulator'] is not None
        )
        data['core-forced'] = (
            'core' in global_settings
            or 'core' in system_settings
            or 'core' in game_settings
            or options['core'] is not None
        )

        if options['emulator'] is not None:
            data['emulator'] = options['emulator']

        if options['core'] is not None:
            data['core'] = options['core']

        if 'use_guns' not in data and options['lightgun']:
            data['use_guns'] = True

        if 'use_wheels' not in data and options['wheel']:
            data['use_wheels'] = True

        # network options
        if options['netplaymode'] is not None:
            data['netplay.mode'] = options['netplaymode']

        if options['netplaypass'] is not None:
            data['netplay.password'] = options['netplaypass']

        if options['netplayip'] is not None:
            data['netplay.server.ip'] = options['netplayip']

        if options['netplayport'] is not None:
            data['netplay.server.port'] = options['netplayport']

        if options['netplaysession'] is not None:
            data['netplay.server.session'] = options['netplaysession']

        # autosave arguments
        if options['state_slot'] is not None:
            data['state_slot'] = options['state_slot']

        if options['autosave'] is not None:
            data['autosave'] = options['autosave']

        if options['state_filename'] is not None:
            data['state_filename'] = options['state_filename']

        return cls(data)


@dataclass
class RenderConfig(Config):
    @classmethod
    def load(cls, options: Options, settings: UnixSettings, shader_set: str | MissingType, /) -> Self:
        data: dict[str, Any] = {}

        if shader_set is not cls.MISSING:
            if shader_set == 'none':
                rendering_defaults = BATOCERA_SHADERS / 'configs' / 'rendering-defaults.yml'
            else:
                rendering_defaults = USER_SHADERS / 'configs' / shader_set / 'rendering-defaults.yml'
                if not rendering_defaults.exists():
                    rendering_defaults = BATOCERA_SHADERS / 'configs' / shader_set / 'rendering-defaults.yml'

            data = _load_defaults(
                options['system'], rendering_defaults, rendering_defaults.with_name('rendering-defaults-arch.yml')
            )

        rom = options['rom']
        rom_settings_name = _get_rom_settings_name(rom)

        # es only allow to update system and game settings in fact for the moment

        # for compatibility with earlier Batocera versions, let's keep -renderer
        # but it should be reviewed when we refactor configgen (to Python3?)
        # so that we can fetch them from system.shader without -renderer
        data.update(settings.get_all_iter(f'{options["system"]}-renderer'))
        data.update(settings.get_all_iter(f'{options["system"]}["{rom_settings_name}"]-renderer'))

        return cls(data)
