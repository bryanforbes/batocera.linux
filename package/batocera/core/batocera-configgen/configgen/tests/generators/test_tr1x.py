from __future__ import annotations

import io
from pathlib import Path
from typing import TYPE_CHECKING
from zipfile import BadZipFile, ZipFile

import pytest

from configgen.batoceraPaths import ROMS
from configgen.config import SystemConfig
from configgen.generators.tr1x.tr1xGenerator import TR1XGenerator
from configgen.utils.download import DownloadException
from tests.generators.base import GeneratorBaseTest

if TYPE_CHECKING:
    from unittest.mock import Mock

    from pyfakefs.fake_filesystem import FakeFilesystem
    from pytest_mock import MockerFixture
    from syrupy.assertion import SnapshotAssertion

    from configgen.controller import ControllerList
    from configgen.Emulator import Emulator
    from configgen.types import Resolution


class TestTR1XGenerator(GeneratorBaseTest):
    @pytest.fixture
    def generator_cls(self) -> type[TR1XGenerator]:
        return TR1XGenerator

    @pytest.fixture
    def system_name(self) -> str:
        return 'traider1'

    @pytest.fixture
    def emulator(self) -> str:
        return 'tr1x'

    @pytest.fixture(autouse=True)
    def fs(self, fs: FakeFilesystem) -> FakeFilesystem:
        fs.create_file(ROMS / 'traider1' / 'Tomb Raider.croft')
        fs.create_file('/usr/bin/tr1x/TR1X', contents='tr1x bin')
        fs.create_file('/usr/bin/tr1x/foo/bar', contents='bar')
        fs.create_file('/usr/bin/tr1x/foo/ham/spam', contents='spam')
        return fs

    @pytest.fixture(autouse=True)
    def download(self, mocker: MockerFixture) -> Mock:
        def side_effect(url: str, _: Path) -> Mock:
            zip_content = io.BytesIO()

            with ZipFile(zip_content, 'w') as zip:
                if url.endswith('/music.zip'):
                    zip.writestr('music/foo.txt', '')
                elif url.endswith('/trub-music.zip'):
                    zip.writestr('bar.txt', '')
                    zip.writestr('DATA/', '')
                    zip.writestr('DATA/CAT.PHD', '')

            zip_content.seek(0)

            mock = mocker.MagicMock()
            mock.__enter__.return_value = zip_content

            return mock

        download = mocker.patch('configgen.generators.tr1x.tr1xGenerator.download')
        download.side_effect = side_effect

        return download

    @pytest.fixture
    def copytree(self, mocker: MockerFixture) -> Mock:
        return mocker.patch('shutil.copytree')

    @pytest.mark.parametrize(
        ('resolution', 'result'),
        [
            ({'width': 640, 'height': 480}, 4 / 3),
            ({'width': 1920, 'height': 1080}, 16 / 9),
            ({'width': 1920, 'height': 1144}, 16 / 9),
            ({'width': 1920, 'height': 1145}, 4 / 3),
        ],
        ids=str,
    )
    def test_get_in_game_ratio(  # pyright: ignore
        self, generator: TR1XGenerator, resolution: Resolution, result: bool
    ) -> None:
        assert generator.getInGameRatio(SystemConfig({}), resolution, Path()) == result

    def test_generate(
        self,
        generator: TR1XGenerator,
        mock_system: Emulator,
        download: Mock,
        one_player_controllers: ControllerList,
        snapshot: SnapshotAssertion,
    ) -> None:
        assert (
            generator.generate(
                mock_system,
                ROMS / 'traider1' / 'Tomb Raider.croft',
                one_player_controllers,
                {},
                [],
                {},
                {'width': 1920, 'height': 1080},
            )
            == snapshot
        )
        assert (ROMS / 'traider1' / 'music' / 'foo.txt').exists()
        assert (ROMS / 'traider1' / 'TR1X').read_text() == 'tr1x bin'
        assert (ROMS / 'traider1' / 'foo' / 'bar').read_text() == 'bar'
        assert (ROMS / 'traider1' / 'foo' / 'ham' / 'spam').read_text() == 'spam'
        download.assert_called_once_with('https://lostartefacts.dev/aux/tr1x/music.zip', ROMS / 'traider1')

    def test_generate_second_run(
        self,
        generator: TR1XGenerator,
        fs: FakeFilesystem,
        mock_system: Emulator,
        download: Mock,
        one_player_controllers: ControllerList,
        snapshot: SnapshotAssertion,
    ) -> None:
        fs.create_dir(ROMS / 'traider1' / 'music')
        fs.create_file(ROMS / 'traider1' / 'TRX1', contents='old tr1x bin')
        fs.create_file(ROMS / 'traider1' / 'foo' / 'bar', contents='old bar')
        fs.create_file(ROMS / 'traider1' / 'foo' / 'ham' / 'spam', contents='old spam')

        assert (
            generator.generate(
                mock_system,
                ROMS / 'traider1' / 'Tomb Raider.croft',
                one_player_controllers,
                {},
                [],
                {},
                {'width': 1920, 'height': 1080},
            )
            == snapshot
        )
        assert (ROMS / 'traider1' / 'TR1X').read_text() == 'tr1x bin'
        assert (ROMS / 'traider1' / 'foo' / 'bar').read_text() == 'bar'
        assert (ROMS / 'traider1' / 'foo' / 'ham' / 'spam').read_text() == 'spam'
        download.assert_not_called()

    @pytest.mark.parametrize('exception', [PermissionError(), Exception()])
    def test_generate_ignores_copy_errors(
        self,
        generator: TR1XGenerator,
        mock_system: Emulator,
        copytree: Mock,
        exception: Exception,
        one_player_controllers: ControllerList,
        snapshot: SnapshotAssertion,
    ) -> None:
        copytree.side_effect = exception

        assert (
            generator.generate(
                mock_system,
                ROMS / 'traider1' / 'Tomb Raider.croft',
                one_player_controllers,
                {},
                [],
                {},
                {'width': 1920, 'height': 1080},
            )
            == snapshot
        )

    @pytest.mark.parametrize('exception', [DownloadException(), BadZipFile(), Exception()])
    def test_generate_ignores_music_download_errors(
        self,
        generator: TR1XGenerator,
        mock_system: Emulator,
        download: Mock,
        exception: Exception,
        one_player_controllers: ControllerList,
        snapshot: SnapshotAssertion,
    ) -> None:
        download.side_effect = exception

        assert (
            generator.generate(
                mock_system,
                ROMS / 'traider1' / 'Tomb Raider.croft',
                one_player_controllers,
                {},
                [],
                {},
                {'width': 1920, 'height': 1080},
            )
            == snapshot
        )

    @pytest.mark.mock_system_config({'tr1x-expansion': '1'})
    def test_generate_expansion(
        self,
        mocker: MockerFixture,
        generator: TR1XGenerator,
        mock_system: Emulator,
        download: Mock,
        one_player_controllers: ControllerList,
        snapshot: SnapshotAssertion,
    ) -> None:
        assert (
            generator.generate(
                mock_system,
                ROMS / 'traider1' / 'Tomb Raider.croft',
                one_player_controllers,
                {},
                [],
                {},
                {'width': 1920, 'height': 1080},
            )
            == snapshot
        )

        assert (ROMS / 'traider1' / 'music' / 'foo.txt').exists()
        assert (ROMS / 'traider1' / 'data' / 'CAT.PHD').exists()
        assert (ROMS / 'traider1' / 'data' / 'bar.txt').exists()

        assert download.call_args_list == [
            mocker.call('https://lostartefacts.dev/aux/tr1x/music.zip', ROMS / 'traider1'),
            mocker.call('https://lostartefacts.dev/aux/tr1x/trub-music.zip', ROMS / 'traider1'),
        ]

    @pytest.mark.mock_system_config({'tr1x-expansion': '1'})
    def test_generate_expansion_downloaded(
        self,
        generator: TR1XGenerator,
        fs: FakeFilesystem,
        mock_system: Emulator,
        download: Mock,
        one_player_controllers: ControllerList,
        snapshot: SnapshotAssertion,
    ) -> None:
        fs.create_dir(ROMS / 'traider1' / 'music')
        fs.create_file(ROMS / 'traider1' / 'data' / 'CAT.PHD')

        assert (
            generator.generate(
                mock_system,
                ROMS / 'traider1' / 'Tomb Raider.croft',
                one_player_controllers,
                {},
                [],
                {},
                {'width': 1920, 'height': 1080},
            )
            == snapshot
        )
        download.assert_not_called()

    @pytest.mark.parametrize('exception', [DownloadException(), BadZipFile(), Exception()])
    @pytest.mark.mock_system_config({'tr1x-expansion': '1'})
    def test_generate_ignores_expansion_download_errors(
        self,
        generator: TR1XGenerator,
        mock_system: Emulator,
        download: Mock,
        exception: type[Exception],
        one_player_controllers: ControllerList,
        snapshot: SnapshotAssertion,
    ) -> None:
        download.side_effect = exception

        assert (
            generator.generate(
                mock_system,
                ROMS / 'traider1' / 'Tomb Raider.croft',
                one_player_controllers,
                {},
                [],
                {},
                {'width': 1920, 'height': 1080},
            )
            == snapshot
        )
