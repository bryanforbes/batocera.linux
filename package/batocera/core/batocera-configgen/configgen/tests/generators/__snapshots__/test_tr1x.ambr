# serializer version: 1
# name: TestTR1XGenerator.test_generate
  tuple(
    list([
      '/userdata/roms/traider1/TR1X',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
    }),
  )
# ---
# name: TestTR1XGenerator.test_generate_expansion
  tuple(
    list([
      '/userdata/roms/traider1/TR1X',
      '-gold',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
    }),
  )
# ---
# name: TestTR1XGenerator.test_generate_expansion_downloaded
  tuple(
    list([
      '/userdata/roms/traider1/TR1X',
      '-gold',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
    }),
  )
# ---
# name: TestTR1XGenerator.test_generate_ignores_copy_errors[exception0]
  tuple(
    list([
      '/userdata/roms/traider1/TR1X',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
    }),
  )
# ---
# name: TestTR1XGenerator.test_generate_ignores_copy_errors[exception1]
  tuple(
    list([
      '/userdata/roms/traider1/TR1X',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
    }),
  )
# ---
# name: TestTR1XGenerator.test_generate_ignores_expansion_download_errors[exception0]
  tuple(
    list([
      '/userdata/roms/traider1/TR1X',
      '-gold',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
    }),
  )
# ---
# name: TestTR1XGenerator.test_generate_ignores_expansion_download_errors[exception1]
  tuple(
    list([
      '/userdata/roms/traider1/TR1X',
      '-gold',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
    }),
  )
# ---
# name: TestTR1XGenerator.test_generate_ignores_expansion_download_errors[exception2]
  tuple(
    list([
      '/userdata/roms/traider1/TR1X',
      '-gold',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
    }),
  )
# ---
# name: TestTR1XGenerator.test_generate_ignores_music_download_errors[exception0]
  tuple(
    list([
      '/userdata/roms/traider1/TR1X',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
    }),
  )
# ---
# name: TestTR1XGenerator.test_generate_ignores_music_download_errors[exception1]
  tuple(
    list([
      '/userdata/roms/traider1/TR1X',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
    }),
  )
# ---
# name: TestTR1XGenerator.test_generate_ignores_music_download_errors[exception2]
  tuple(
    list([
      '/userdata/roms/traider1/TR1X',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
    }),
  )
# ---
# name: TestTR1XGenerator.test_generate_second_run
  tuple(
    list([
      '/userdata/roms/traider1/TR1X',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'SDL_JOYSTICK_HIDAPI': '0',
    }),
  )
# ---
# name: TestTR1XGenerator.test_get_hotkeys_context
  dict({
    'keys': dict({
      'exit': list([
        'KEY_LEFTALT',
        'KEY_F4',
      ]),
      'restore_state': 'KEY_F6',
      'save_state': 'KEY_F5',
    }),
    'name': 'tr1x',
  })
# ---
